catch JS exceptions when using InvokeAsync and throw them in .NET

campersau opened the pull request:
# catch JS exceptions when using InvokeAsync and throw them in .NET

Exceptions in `createJSCallResult` and `stringifyArgs` were only thrown in JS before and .NET never received them.

Examples of exceptions which are now reaching .NET:
- https://github.com/dotnet/aspnetcore/blob/371bf375f5bf6e48ab6b33e24ad0c73a940b4ae4/src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts#L591
- https://github.com/dotnet/aspnetcore/blob/371bf375f5bf6e48ab6b33e24ad0c73a940b4ae4/src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts#L167-L188
- https://github.com/dotnet/aspnetcore/blob/371bf375f5bf6e48ab6b33e24ad0c73a940b4ae4/src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts#L148

Fixes #40272 (references the last exception from the list above)
Fixes #41934 (exception in `stringifyArgs`)

campersau commented on src/Components/WebAssembly/WebAssembly/test/JSObjectReferenceJsonConverterTest.cs:
```diff
@@ -17,6 +17,20 @@ public JSObjectReferenceJsonConverterTest()
     }
 
+    [Fact]
+    public void Read_ReadsJson_IJSObjectReference()
```
Unrelated bonus test.

mkArtakMSFT commented:
Thanks for your PR, @campersau.
@MackinnonBuck can you please review this? Thanks!

campersau committed:
catch JS exceptions in InvokeAsync and throw them in .NET

campersau commented:
Updated to also catch exceptions during `stringifyArgs` which is the case in https://github.com/dotnet/aspnetcore/issues/41934 and added tests for synchronus `Invoke` to show that the behavior is the same as `InvokeAsync`, both throwing exceptions.

MackinnonBuck reviewed (approved):
LGTM

japui commented:
catch JS exceptions when using InvokeAsync and throw them in .NET

MackinnonBuck commented:
Thanks @campersau!

datvm commented:
Hi is this PR actually online yet? I am using .NET 7 Blazor WebAssembly (7.0.202) and I still have this issue. The code runs fine if JS returns something but does not returns when JS returns `undefined` (Error printed in Console but nothing happens on .NET side).

```cs
    public static async Task<T> GetGlobalObject<T>(this IJSRuntime js, string key) =>
        await js.InvokeAsync<T>(
            GetFullFuncName("getGlobalObj"),
            key);

    public static async Task<IJsWindowReference?> GetOpenerAsync(this IJSRuntime js)
    {
        var opener = await js.GetGlobalObject<IJSObjectReference?>("opener");
        return opener?.AsWindow(js);
    }
```

```js
    getGlobalObj(key) {
        return globalThis[key];
    }
```
